pyro/distributions/householder.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import absolute_import, division, print_function
+
+import math
+import warnings
+
+import torch
+import torch.nn as nn
+from torch.distributions import constraints
+
+from pyro.distributions.torch_transform import TransformModule
+from pyro.distributions.util import copy_docs_from
+
+
+@copy_docs_from(TransformModule)
+class HouseholderFlow(TransformModule):
+    """
+    A single transformation of Householder flow,
+
+        :math:`\\mathbf{y} = (I - 2*\\frac{\\mathbf{u}\\mathbf{u}^T}{||\\mathbf{u}||^2})\\mathbf{x}`
+
+    where :math:`\\mathbf{x}` are the inputs, :math:`\\mathbf{y}` are the outputs, and the learnable parameters
+    are :math:`\\mathbf{u}\\in\\mathbb{R}^D` for input dimension :math:`D`.
+
+    The transformation represents the reflection of :math:`\\mathbf{x}` through the plane passing through the
+    origin with normal :math:`\\mathbf{u}`. Together with `TransformedDistribution` this provides a way to
+    create richer variational approximations.
+
+    :math:`D` applications of this transformation are able to transform standard i.i.d. standard Gaussian noise
+    into a Gaussian variable with an arbitrary covariance matrix. With :math:`K<D` transformations, one is able
+    to approximate a full-rank covariance Gaussian distribution.
+
+    Example usage:
+
+    >>> base_dist = dist.Normal(torch.zeros(10), torch.ones(10))
+    >>> flows = [HouseholderFlow(10) for _ in range(3)]
+    >>> [pyro.module("my_flow", p) for f in flows] # doctest: +SKIP
+    >>> flow_dist = dist.TransformedDistribution(base_dist, flows)
+    >>> flow_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    :param input_dim: the dimension of the input (and output) variable.
+    :type input_dim: int
+    :param count_transforms: number of layers of Householder flow to apply.
+    :type count_transforms: int
+
+    References:
+
+    Improving Variational Auto-Encoders using Householder Flow, [arXiv:1611.09630]
+    Tomczak, J. M., & Welling, M.
+
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 1
+
+    def __init__(self, input_dim, count_transforms=1):
+        super(HouseholderFlow, self).__init__(cache_size=1)
+
+        self.input_dim = input_dim
+        assert count_transforms > 0
+        if count_transforms > input_dim:
+            warnings.warn(
+                "{} layers of Householder flow applied to input of dimension {}".format(
```

Comment:
explain why this is unnecessary?

Reply:
I put this in because I don't think there is an advantage to multiplying more than D Householder matrices for input dim D. Full-rank is as high as you can go!

Reply:
Ah I see you meant that I should add in an explanatory note to the user. I've done this now!